water_neuron/src/lib.rs

```diff
@@ -149,6 +153,89 @@ pub struct CanisterInfo {
     pub governance_fee_share_percent: u64,
 }
 
+#[derive(Debug, EnumString, EnumIter, Display)]
+pub enum Icrc21Function {
+    #[strum(serialize = "icp_to_nicp")]
+    Stake,
+    #[strum(serialize = "nicp_to_icp")]
+    Unstake,
+    #[strum(serialize = "cancel_withdrawal")]
+    CancelWithdrawal,
+    #[strum(serialize = "claim_airdrop")]
+    ClaimAirdrop,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct StandardRecord {
+    pub url: String,
+    pub name: String,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ConsentMessageRequest {
+    pub arg: Vec<u8>,
+    pub method: String,
+    pub user_preferences: ConsentMessageSpec,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ConsentMessageSpec {
+    pub metadata: ConsentMessageMetadata,
+    pub device_spec: Option<DisplayMessageType>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ConsentMessageMetadata {
+    pub utc_offset_minutes: Option<u16>,
+    pub language: String,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum DisplayMessageType {
+    GenericDisplay,
+    LineDisplay {
+        characters_per_line: u16,
+        lines_per_page: u16,
+    },
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ConsentInfo {
+    pub metadata: ConsentMessageMetadata,
+    pub consent_message: ConsentMessage,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum ConsentMessage {
+    LineDisplayMessage { pages: Vec<LineDisplayPage> },
+    GenericDisplayMessage(String),
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct LineDisplayPage {
+    lines: Vec<String>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum Icrc21Error {
+    GenericError {
+        description: String,
+        error_code: u64,
+    },
+    InsufficientPayment(ErrorInfo),
+    UnsupportedCanisterCall(ErrorInfo),
+    ConsentMessageUnavailable(ErrorInfo),
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ErrorInfo {
+    pub description: String,
+}
+
+pub fn convert_amount_e8s_to_string_representation(amount_e8s: u64, decimals: i32) -> String {
```

Comment:
This is wrong, look at DisplayAmount in `dashboard.rs`. float are rarely good for anything.